pkg/actions/db_service.go

```diff
@@ -402,3 +428,185 @@ func (s *DBService) NewRunID() string {
+				PartitionKey: []byte(PartitionKey),
+				Key:          secKey,
+				Value:        data,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// tasksWorker dispatch taskWorker jobs per run
+func tasksWorker(ctx context.Context, d *pgxpool.Pool, jobChan <-chan *migrateTask) error {
+	tasksChan := make(chan *taskResultMigrate, migrateQueueSize)
+	var g multierror.Group
+	for i := 0; i < taskWorkers; i++ {
+		g.Go(func() error {
+			return taskWorker(tasksChan)
+		})
+	}
+	for t := range jobChan {
+		tasksRows, err := d.Query(ctx, "SELECT * FROM actions_run_hooks where run_id=$1", t.oldID)
+		if err != nil {
+			return err
+		}
+		taskScanner := pgxscan.NewRowScanner(tasksRows)
+		for tasksRows.Next() {
+			tm := new(taskResultMigrate)
+			err = taskScanner.Scan(tm)
+			if err != nil {
+				return err
+			}
+			tm.RunID = t.newID
+			tasksChan <- tm
+		}
+		tasksRows.Close()
+	}
+	close(tasksChan)
+
+	return g.Wait().ErrorOrNil()
+}
+
+// taskWorker writing task data to fd
+func taskWorker(jobChan <-chan *taskResultMigrate) error {
+	for t := range jobChan {
+		taskPr := protoFromTaskResult(&t.TaskResult)
+		taskValue, err := proto.Marshal(taskPr)
+		if err != nil {
+			return err
+		}
+		taskKey := []byte(kv.FormatPath(TasksPath(t.RepoID, t.RunID), t.HookRunID))
+		if err = encoder.Encode(kv.Entry{
+			PartitionKey: []byte(PartitionKey),
+			Key:          taskKey,
+			Value:        taskValue,
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Migrate(ctx context.Context, d *pgxpool.Pool, writer io.Writer) error {
+	encoder = kv.SafeEncoder{
+		Je: json.NewEncoder(writer),
+		Mu: sync.Mutex{},
+	}
+	// Create header
+	if err := encoder.Encode(kv.Header{
+		LakeFSVersion:   version.Version,
+		PackageName:     packageName,
+		DBSchemaVersion: kv.InitialMigrateVersion,
+		CreatedAt:       time.Now().UTC(),
+	}); err != nil {
+		return err
+	}
+
+	runChan := make(chan *runResultMigrate, migrateQueueSize)
+	tasksChan := make(chan *migrateTask, migrateQueueSize)
+	var g multierror.Group
+	for i := 0; i < runWorkers; i++ {
+		// Run worker reads runs from DB and writes respected entries to fd
+		g.Go(func() error {
+			return runWorker(runChan)
+		})
+		// tasks worker scans tasks per run ID and starts the task go routines which write the respected entries to fd
+		g.Go(func() error {
+			return tasksWorker(ctx, d, tasksChan)
+		})
+	}
+
+	rows, err := d.Query(ctx, "SELECT * FROM actions_runs ORDER BY run_id DESC")
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+	rowScanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		m := new(runResultMigrate)
+		err = rowScanner.Scan(m)
+		if err != nil {
+			return err
+		}
+		runTime, err := time.Parse(graveler.RunIDTimeLayout, m.RunResult.RunID[:len(graveler.RunIDTimeLayout)])
+		if err != nil {
+			return err
+		}
+		newRunID := newRunIDFromTime(runTime)
+		oldRunID := m.RunResult.RunID
+		m.RunResult.RunID = newRunID
+		runChan <- m
+		t := &migrateTask{
```

Comment:
Is there any added value to have the tasks migration dependant on the corresponding run migration?
What I'm asking is, if we had 2 separate iterations, one over `action_runs` and the other over `actions_run_hooks` (regardless of `run_id`), having each trigger a job for each row, will it make any difference? (Other than the need to recalculate `newRunID` potentially many times)
The reason I'm asking is that if it is the same, each of these loops will look pretty similar to the `mulriparts.Migrate` loop and that makes me think of a generic `Migrate` that gets the specifics as paramaters (DB table name, type for marshaling etc.) and performs the "same" work for all
I guess this is probably not the right place to make such a change, but weas it considered?

Reply:
That's a great point. At the beginning the new run ID was created independently from the original ID so we had to save the tasks in the context of the run. We can definitely do it as you suggest and check if there's an improvement on performance. Regarding your second thought - I think this could be quite more tricky since the migration operation doesn't involve just transposing the data from the DB world to the KV. It also requires creating secondary indices and the conversion of some of the data (as in the example of actions and auth).